chainstate/src/detail/tests/normal_tokens.rs

```diff
@@ -0,0 +1,1650 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::detail::tests::test_framework::{TestFramework, TransactionBuilder};
+use crate::detail::transaction_verifier::error::ConnectTransactionError;
+use crate::detail::CheckBlockTransactionsError;
+use crate::{
+    detail::{tests::TestBlockInfo, CheckBlockError, TokensError},
+    BlockError, BlockSource,
+};
+use common::primitives::{id, Id};
+use common::{
+    chain::{
+        signature::inputsig::InputWitness,
+        tokens::{token_id, OutputValue, TokenData, TokenId},
+        Destination, OutputPurpose, TxInput, TxOutput,
+    },
+    primitives::{Amount, Idable},
+};
+use crypto::{hash::StreamHasher, random::Rng};
+use expect_test::expect;
+use rand::distributions::uniform::SampleRange;
+use rstest::rstest;
+use test_utils::random::{make_seedable_rng, Seed};
+
+fn random_string<RNG: Rng, R: SampleRange<usize>>(rng: &mut RNG, range_len: R) -> String {
+    use rand::distributions::{Alphanumeric, DistString};
+    if range_len.is_empty() {
+        return String::new();
+    }
+    let len = rng.gen_range(range_len);
+    Alphanumeric.sample_string(rng, len)
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn token_issue_test(#[case] seed: Seed) {
```

Comment:
@TheQuantumPhysicist what do you think? Some tests required creating a chain, and seems can't be simplified.

Reply:
On one hand, the format of the tests were chaotic when we started the project because the test framework wasn't complete. On the other hand, keep in mind that many tests don't really share the exact same fixture. It's a game of whac-A-mole to hit them all and not make an extended mess with another layer of test-framework that just parametrizes everything in the fixture and just passes it to the test framework.

Reply:
Test fixture shouldn't have more code that it's already written for test initialisation.
My point is that it's better to try to write a lot of small unit test than few big.